File: Data Structures & Algorithms/balanced-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    // {balanced, height}
   vector<int> dfs(TreeNode* root) {
       if (not root) {
          return {1, 0};
       }
       auto left = dfs(root->left);
       auto right = dfs(root->right);
       bool balanced = (left[0] and right[0] and
                        (abs(left[1] - right[1]) < 2));
        int height = 1 + max(left[1], right[1]);
        return {balanced, height};
   }
public:
    bool isBalanced(TreeNode* root) {
        return dfs(root)[0] == 1;
    }
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int MaxPathSum(TreeNode root) {
        int ans = root.val;
        Dfs(root, ref ans);
        return ans;
    }
    int GetMaxPathSum(TreeNode root) {
        if(root is null) return 0;
        int left = GetMaxPathSum(root.left);
        int right = GetMaxPathSum(root.right);
        int sum = root.val + Math.Max(left,right);
        return Math.Max(0, sum);
    }    
    void Dfs(TreeNode root, ref int ans) {
        if (root is null) return;
        int left = GetMaxPathSum(root.left);
        int right = GetMaxPathSum(root.right);
        ans = Math.Max(ans, root.val + left + right);
        Dfs(root.left, ref ans); Dfs(root.right, ref ans);
    }
}
